It's ticking!

BautistaDalmasso opened the pull request:
Server now performs one tick every 6 seconds.
Updates about the players actions are added to a queue.Queue, and then they are sent to each player as necessary.

- [x] Movement
- [x] Fighting 
- [x] Fleeing
- [ ] Dying (both mobs and players)
- [ ] Leveling up (and becoming a butterfly)
- [ ] Entities in the room.
- [ ] Targeted attacks from the client.

BautistaDalmasso committed:
Server now ticks! And sends clients some updates.

Server now performs one tick every 6 seconds.
Updates about the players actions are added to a queue.Queue, and then
they are sent to each player as necessary.

A5rocks commented on server/game_components/game.py:
```diff
@@ -19,6 +22,7 @@ class Game:
     start_time: int
 
     def __init__(self):
+        self.out_queue: queue.Queue[ActionDict | MovementDict | int] = queue.Queue()
```
Should this be an `asyncio.Queue`?

BautistaDalmasso committed:
Fixes some annoyances with imports.

BautistaDalmasso committed:
Allows the server to handle `clear`, `nvm` and `flee` actions.

BautistaDalmasso commented:
whoops, lots of force push's sorry.

BautistaDalmasso committed:
Merge branch 'main' into server-updating

BautistaDalmasso committed:
Merge branch 'main' into server-updating

BautistaDalmasso committed:
Rooms can now send updates to the players in them.

Rooms now track events that should be broadcasted to the players that are
inside said room.
After a game tick these events are sent to the players.

AbooMinister25 commented on server/main.py:
```diff
@@ -153,6 +160,111 @@ async def game_loop():
+                player_connections = {
+                    connections.get(player_uid)
+                    for player_uid in player_uids
+                    if connections.get(player_uid) is not None
```
the `is not None` here isn't necessary, just `if connections.get(player_uid)` is fine.

BautistaDalmasso replied:
True, but it doesn't hurt being a bit more explicit imo.

AbooMinister25 replied:
Sure, but in this case (as with `if truthy` being preferred over `if truthy is True`) it's OK just to leave out `is not None`.

AbooMinister25 commented on server/main.py:
```diff
@@ -153,6 +160,111 @@ async def game_loop():
+
+def get_room_update_message(room_action: RoomActionDict) -> str:
+    """This is the message to be displayed to the players in the action that this room happened."""
+    RoomActionDict()
```
As far as I can tell, this doesn't have any effect. What's it for?

BautistaDalmasso replied:
it doesn't, I put it there because it helped me with some type hints and forgot to remove it :PPP

BautistaDalmasso committed:
Players are now registered as having entered combat.

BautistaDalmasso committed:
Fix bug with event updates "skipping" one tick.

BautistaDalmasso committed:
Small improvement to action messges.

BautistaDalmasso committed:
Improves the way our server handles and responds to actions.

BautistaDalmasso committed:
Player can now flee combat.

BautistaDalmasso committed:
Mobs can now act during combat.

BautistaDalmasso committed:
Players now get notified when someone enters or leaves their room.

BautistaDalmasso committed:
Removes unused class `Combat`.

We ended up using the rooms to handle combat.

AbooMinister25 reviewed (approved):
LGTM
